dlrover/python/elastic_agent/diagnosis/diagnosis_agent.py

```diff
@@ -230,35 +249,27 @@ def _periodically_diagnosis(self):
                 DiagnosisConstant.AGENT_PERIODICALLY_DIAGNOSIS_INTERVAL_SECS
             )
 
-    def diagnose_training_failure(self) -> NodeAction:
+    def diagnose_training_failure(self) -> DiagnosisAction:
         self._report_failure_to_master(
             self._agent_context.run_result.failures,
             self._agent_context.restart_count,
         )
-        # check if the node is failed
-        inference = Inference(
-            name=InferenceName.NODE,
-            attribution=InferenceAttribute.ISORNOT,
-            description=InferenceDescription.FAILURE,
-            configs={
-                InferenceConfigKey.LOG_FILE: self._training_log_file,
-                InferenceConfigKey.ERRORS: self._errors,
-            },
+        ob_action = self._diagnostician_mgr.observe(
+            DiagnosisAgentConstants.NODE_FAILED,
+            log_file=self._training_log_file,
+            errors=self._errors,
         )
-        ic = InferenceChain([inference], get_training_failure_operators())
-        infer_results = ic.infer()
-        failure_inf = Inference(
-            name=InferenceName.NODE,
-            attribution=InferenceAttribute.IS,
-            description=InferenceDescription.FAILURE,
-        )
-        failure_node = is_inference_included(infer_results, failure_inf)
 
-        if self._agent_context.remaining_failovers > 0 and not failure_node:
+        node_failed = False
+        if isinstance(ob_action, ObservationAction):
+            ob_action.__class__ = ObservationAction
+            node_failed = ob_action.node_failed()
+
+        if self._agent_context.remaining_failovers > 0 and not node_failed:
```

Comment:
The following logic should be defined in a 'resolver' of the worker side(difference from master side).